test/extended/dns/dns_libraries.go

```diff
@@ -0,0 +1,134 @@
+			e2e.Failf("DNS resolution failed: %s", line)
+		}
+	}
+}
+
+var _ = Describe("[sig-network-edge] DNS lookup", func() {
+	f := e2e.NewDefaultFramework("dns-libraries")
+	oc := exutil.NewCLI("dns-libraries")
+
+	// creates a simple Pod that is using the image under /testdata/dns, which performs a DNS query to make sure Go's DNS resolver works fine with OpenShift DNS.
+	It("using Go's DNS resolver", func() {
+		configClient, err := configclient.NewForConfig(f.ClientConfig())
+		if err != nil {
+			e2e.Failf("Failed to get config client: %v", err)
+		}
+		infra, err := configClient.ConfigV1().Infrastructures().Get(context.Background(), "cluster", metav1.GetOptions{})
+		if err != nil {
+			e2e.Failf("Failed to get cluster config: %v", err)
+		}
+		if infra.Status.PlatformStatus.Type == configv1.BareMetalPlatformType {
+			Skip("Skip DNS lookup using Go's DNS resolver on BareMetal as there is no external connection to pull the base image.")
+		}
+
+		dnsService, err := f.ClientSet.CoreV1().Services("openshift-dns").Get(context.Background(), "dns-default", metav1.GetOptions{})
+		if err != nil {
+			e2e.Failf("Failed to get default DNS service: %v", err)
+		}
+
+		goVersions := []string{"go-v1-16", "go-v1-17"}
+		for _, version := range goVersions {
+			By(fmt.Sprintf("creating build and deployment config for dns-libraries-%s", version))
+			err = oc.Run("create").Args("-f", exutil.FixturePath("testdata", "dns", version, "dns_libraries_go.yaml")).Execute()
+			if err != nil {
+				e2e.Failf("Failed to create build configs and deployment config for dns-libraries-go: %v", err)
+			}
+
+			By("starting the builder image build with a directory")
+			err = oc.Run("start-build").Args(fmt.Sprintf("dns-libraries-%s", version), fmt.Sprintf("--from-dir=%s", exutil.FixturePath("testdata", "dns", version))).Execute()
+			if err != nil {
+				e2e.Failf("Failed to start the builder image build: %v", err)
+			}
+
+			labels := exutil.ParseLabelsOrDie(fmt.Sprintf("app=dns-libraries-%s", version))
+
+			By(fmt.Sprintf("expect the builds to complete successfully and deploy a dns-libraries-%s pod", version))
+			pods, err := exutil.WaitForPods(oc.KubeClient().CoreV1().Pods(oc.Namespace()), labels, exutil.CheckPodIsRunning, 1, 5*time.Minute)
+			if err != nil {
+				e2e.Failf("Failed to start the dns-libraries-go pod: %v", err)
+			}
+			if len(pods) != 1 {
+				e2e.Failf("Got %d pods with labels %v, expected 1", len(pods), labels)
+			}
+
+			By("expect the go application to successfully resolve DNS")
+			pod, err := oc.KubeClient().CoreV1().Pods(oc.Namespace()).Get(context.Background(), pods[0], metav1.GetOptions{})
+			if err != nil {
+				e2e.Failf("Failed to get dns-libraries-go pod: %v", err)
+			}
+			//execute "go run /go/dns_libraries.go -cluster-ip={CLUSTER_IP}" command inside the pod
+			args := []string{pod.Name, "-c", pod.Spec.Containers[0].Name, "--", "bash", "-c", fmt.Sprintf("go run /go/dns_libraries.go -cluster-ip=%q", dnsService.Spec.ClusterIP)}
```

Comment:
FWIW there is a `RunHostCmdOrDie` helper function that could clean this up a bit.